yeagerai/genlayer-simulator
# 348 sim be add precommit hooks for testing code linting and other conventions

Fixes https://github.com/yeagerai/genlayer-simulator/issues/348

# What



- added `precommit` hooks for linting and backend unit tests 
- added a GH action to run `pre-commit`
  - given that pre-commit is opt-in, I think it's a good idea to check if the conventions are being followed

# Why



- to force developers to apply standards before committing, reducing the time spent on discussions around these details

# Testing done



- I've made commits and I've checked that the pre-commit hooks were working

# Decisions made



- pre-commit needed was added as a dev dependency, it is installed with Python (documented in `Contributing.md`)
- installing npm packages was added as dev dependencies for the frontend commits

The other alternative I thought of would be to use Docker to not use `pre-commit` (or the other dependencies). I didn't think it made sense given that there are few dependencies and they are not system-wide, so it makes sense to install them locally. Also, the time it takes to run them must be quick in order to be commit hooks, which docker would slow down a bit.
This is a decision we can discuss

# Checks

- [x] I have created an issue for this PR
- [x] I have set the PR name to the issue name

# Reviewing tips

a) 
- Many files were modified due to the fixes of the linters (yay it's already working)
- Remove whitespace changes when reviewing

b) Test it out by
- check out this branch
- try following the `Contributing.md` guide to see if it works for you
- try `git commit --empty`
